After talking about the model catalog today, I got curious about the DOIs that have been minted in the USGS ID space that claim to be models. This notebook takes a look. A quick look at the DataCite works API shows that there are quite a few DOIs minted, with the majority coming from work I did to register all those GAP species habitat models. But there seem to be some other interesting things in there. I put in a process to pull all those together by paginating through the API results but then stash them in a Feather file in case for further reference.

In [1]:
import requests
import pandas as pd

recordset = list()

In [2]:
# Function to create lookup URL for USGS DOIs where resource type is model
def get_url(page_number):
    return f"https://api.datacite.org/works?data-center-id=usgs.prod&resource-type-id=model&page[number]={page_number}"


In [3]:
%%time
# Get the first page of results and set up a structure to contain everything
page_number = 1
result = requests.get(get_url(page_number)).json()

CPU times: user 26.7 ms, sys: 7.96 ms, total: 34.6 ms
Wall time: 1.69 s


In [4]:
%%time
# Loop through and build out the full result set (the lazy way)
while len(result["data"]) > 0:
    recordset.extend(result["data"])
    page_number += 1
    result = requests.get(get_url(page_number)).json()

CPU times: user 1.33 s, sys: 104 ms, total: 1.44 s
Wall time: 5min 57s


If we ran the process to get the latest set of model type DOIs, then we build a dataframe from the attributes structure in each record. If we didn't run it, then we pull in the dataframe from the stashed Pickle file.

In [7]:
if len(recordset) == 0:
    df_models = pd.read_pickle("ModelsStash")
else:
    df_models = pd.DataFrame([i["attributes"] for i in recordset])
    df_models.to_pickle("ModelsStash")


In [8]:
# See what the data look like in a dataframe
df_models.head(5)

doi                        identifier  \
0  10.5066/p9hiyvg2  https://doi.org/10.5066/p9hiyvg2   
1  10.5066/f76t0jpb  https://doi.org/10.5066/f76t0jpb   
2  10.5066/f798853m  https://doi.org/10.5066/f798853m   
3  10.5066/f76w98jb  https://doi.org/10.5066/f76w98jb   
4  10.5066/f7h99392  https://doi.org/10.5066/f7h99392   

                                                 url  \
0  https://water.usgs.gov/GIS/metadata/usgswrd/XM...   
1  https://ca.water.usgs.gov/projects/reg_hydro/b...   
2  http://water.usgs.gov/GIS/metadata/usgswrd/XML...   
3  https://water.usgs.gov/GIS/metadata/usgswrd/XM...   
4  http://water.usgs.gov/GIS/metadata/usgswrd/XML...   

                                              author  \
0       [{'given': 'Tracie R', 'family': 'Jackson'}]   
1  [{'given': 'Lorraine E.', 'family': 'Flint'}, ...   
2  [{'given': 'J Hal', 'family': 'Davis'}, {'give...   
3  [{'given': 'Alex R', 'family': 'Fiore'}, {'giv...   
4  [{'given': 'Stephen J', 'family': 'Cauller'}, ...   

                                               title         container-title  \
0  MODFLOW-2005 and PEST models used to character...  U.S. Geological Survey   
1  California Basin Characterization Model: A Dat...  U.S. Geological Survey   
2  MODFLOW 2000 and MT3DMS models of potentiometr...  U.S. Geological Survey   
3  MODFLOW-2005 model used to evaluate the potent...  U.S. Geological Survey   
4  MODFLOW2005 model used to simulate the effects...  U.S. Geological Survey   

                                         description resource-type-subtype  \
0                                               None                 Model   
1  The Basin Characterization Model (BCM), can tr...                 Model   
2  This model is a preliminary characterization o...                 Model   
3  A three-dimensional groundwater flow model, MO...                 Model   
4  A three-dimensional groundwater flow model was...                 Model   

  data-center-id member-id  ... view-count views-over-time download-count  \
0      usgs.prod      usgs  ...          0              []              0   
1      usgs.prod      usgs  ...          0              []              0   
2      usgs.prod      usgs  ...          0              []              0   
3      usgs.prod      usgs  ...          0              []              0   
4      usgs.prod      usgs  ...          0              []              0   

  downloads-over-time published                registered  checked  \
0                  []      2020  2020-03-05T18:30:50.000Z     None   
1                  []      2014  2014-09-17T00:57:18.000Z     None   
2                  []      2016  2016-05-05T18:15:15.000Z     None   
3                  []      2018  2018-03-13T14:32:08.000Z     None   
4                  []      2016  2016-07-26T16:40:24.000Z     None   

                    updated  media  \
0  2020-03-05T18:30:50.000Z     []   
1  2020-03-02T21:34:34.000Z     []   
2  2020-03-02T21:18:55.000Z     []   
3  2020-03-02T21:18:05.000Z     []   
4  2020-03-02T21:16:07.000Z     []   

                                                 xml  
0  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
1  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
2  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
3  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
4  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  

[5 rows x 28 columns]

In [9]:
# Set the option to be able to look at everything in columns for convenience
pd.set_option('display.max_colwidth', None)

One of the things I was interested in is where these DOIs actually point in terms of their dereferencing URLs. To examine this, I create a new column with just the domain part of the URL so that I can group and look at things.

In [10]:
df_models['url_domain'] = df_models.apply(lambda row: row.url.split("/")[2], axis = 1)

In [11]:
df_models[["doi","url_domain"]].groupby('url_domain').count()

doi
url_domain                        
alaska.usgs.gov                  1
axiomdatascience.com             1
ca.water.usgs.gov                1
coastal.er.usgs.gov              1
code.usgs.gov                    2
dx.doi.org                       1
github.com                       1
my.usgs.gov                      3
nrtwq.usgs.gov                   3
regclim.coas.oregonstate.edu     1
usgs.gov                         2
walrus.wr.usgs.gov               1
water.usgs.gov                  93
www.mbr-pwrc.usgs.gov            2
www.sciencebase.gov           1761
www.usgs.gov                     2

That's kind of an interesting spread. Almost all of the 1761 "models" in ScienceBase are the GAP habitat maps, which are a model output of sorts from a habitat affinity-based species distribution modeling method that involves human intervention. But it looks like we have some other interesting things to look at and think about in terms of what they actual represent for a future model catalog. In the sections below, I output particular records, take a look to see what's on the other end, and provide some notes.

In [13]:
df_models.loc[df_models['url_domain'] == "usgs.gov"][["doi","url","title","description"]]

doi               url  title                 description
1874  10.5066/f78050nh   http://usgs.gov  Title                  test111222
1875  10.5066/f7cr5rd0  https://usgs.gov  Title  description of the dataset

Well, those are disappointing. Looks like someone needs to do some cleanup in what actually got registered out with DataCite, if that's even possible.

In [12]:
df_models.loc[df_models['url_domain'] == "axiomdatascience.com"][["doi","url","title","description"]]

doi  \
113  10.5066/f72b8w3t   

                                                            url  \
113  http://axiomdatascience.com/maps/integrated/?portal_id=51#   

                                                                                                                           title  \
113  Wave and Wind projections for United States Coasts; Mainland, Pacific Islands, and United States-Affiliated Pacific Islands   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               

This one from a USGS partner in Alaska that I know pretty well looks interesting. The DOI de-references to a web application run by Axiom Data Science that provides an interactive system to visualize and download the cached results (model output data) for a wave height and wind speed model that looks at both historic conditions and projected futures under several scenarios for coastal zones. The abstract and the web site list other interesting parts of the picture, including a USGS Open File Report describing the methods and what is possibly an "off the books" web page that purportedly accompanied a data release but is the kind of thing that might otherwise have ended up in a USGS Data Series or something similar with a journal. This is a pretty interesting case that has just about all the components for a model catalog except that I couldn't find the source code with a cursory look at least.

In [14]:
df_models.loc[df_models['url_domain'] == "code.usgs.gov"][["doi","url","title","description"]]

doi  \
344  10.5066/p9nquaow   
481  10.5066/p9nqnh41   

                                                               url  \
344                       https://code.usgs.gov/coawstmodel/COAWST   
481  https://code.usgs.gov/ecosystems/FRESC/sagebrush_hurdle_model   

                           title  \
344  COAWST Modeling System v3.4   
481       sagebrush_hurdle_model   

                                                          description  
344  Coupled ocean atmosphere wave sediment transport modeling system  
481                                                              None